### PR TITLE
Circular gradient

### PR DESCRIPTION
**Pull request by niravasher**
Fixes #[Add issue number here.]

Make sure these boxes are checked before your pull request (PR) is ready to be reviewed and merged. Thanks!

* [x] tests pass -- look for a green checkbox ✔️ a few minutes after opening your PR -- or run tests locally with `npm test`
* [x] code is in uniquely-named feature branch and has no merge conflicts
* [x] PR is descriptively titled
* [x] ask `@publiclab/is-reviewers` for help, in a comment below

> We're happy to help you get this ready -- don't be afraid to ask for help, and **don't be discouraged** if your tests fail at first!

If tests do fail, click on the red `X` to learn why by reading the logs.

Please be sure you've reviewed our contribution guidelines at https://publiclab.org/contributing-to-public-lab-software

Thanks!

### CODE REVIEW

**Comment by niravasher** [comment]
@publiclab/is-reviewers please have a look at it

**Comment by HarshKhandeparkar** [comment]
@publiclab/is-reviewers

File: src/modules/CircularGradient/Module.js [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+            var totalLength = 0;
+            var r = savePixels(pixels, input.format, { quality: 100 });
+
+            r.on("data", function(chunk) {
+                totalLength += chunk.length;
+                chunks.push(chunk);
+            });
+
+            r.on("end", function() {
+                var data = Buffer.concat(chunks, totalLength).toString("base64");
```
Comment (HarshKhandeparkar) [inline src/modules/CircularGradient/Module.js]: You don't have to do the pixel saving yourself. PixelManipulation will do it for you.

**Comment by HarshKhandeparkar** [comment]
The best example for a module is drawRectangle

https://github.com/publiclab/image-sequencer/blob/main/src/modules/DrawRectangle/moduke.js
And

https://github.com/publiclab/image-sequencer/blob/main/src/modules/DrawRectangle/DrawRectangle.js

**Comment by niravasher** [comment]
I hope the PR got updated. Also keeping the save function for now to test something, will remove later

**Comment by niravasher** [comment]
The putput is a weird looking gradient in the top-left quadrant. If I include the top-left quadrant, it just overlaps the top left quadrant with a linear gradient
Screenshot (42) @HarshKhandeparkar

File: src/modules/CircularGradient/Module.js [hunk 2]
```diff
@@ -28,42 +34,40 @@ module.exports = function Circular(options, UI) {
...
-                decrease = 255;
-                for (var j = 0 ; j < height/2-1 ; j++) {
-                    pixels.set(i, j, 0, 0);
-                    pixels.set(i, j, 1, 0);
-                    pixels.set(i, j, 2, 0);
-                    pixels.set(i, j, 3, decrease);
-                    decrease-=decrement;
-                }
-                decrement+=0.2;
+            if (options.r > 210) {
```
Comment (jywarren) [inline src/modules/CircularGradient/Module.js]: So here, you could remove all the `options.r`, `options.g`, and `options.b`...

File: src/modules/CircularGradient/Module.js [hunk 3]
```diff
@@ -28,42 +34,40 @@ module.exports = function Circular(options, UI) {
...
-                    pixels.set(i, j, 3, 255-decrease);
-                    decrease-=decrement;
+            for (var i = 0 ; i < width ; i++) {
+                for (var j = 0 ; j < height ; j++) {
+                    var distX = Math.abs(options.centreX - i);
+                    var distY = Math.abs(options.centreY - j);
+                    var distance = Math.sqrt(Math.pow(distX, 2) + Math.pow(distY, 2));
+                    var brightness = 255 * (distance / (2*width));
+                    if (distance-1 > width) {distance = width;}
+                    pixels.set(i, j, 0, options.r*1.4*distance/width);
```
Comment (jywarren) [inline src/modules/CircularGradient/Module.js]: Then here, you'd just do distance/width, without using a color, and the same for the next 2 lines.

File: src/modules/CircularGradient/info.json [hunk 4]
```diff
@@ -2,7 +2,36 @@
     "name": "Circular Gradient",
     "description": "Gives a circular gradient of the image",
     "inputs": {
-
+        "r": {
```
Comment (jywarren) [inline src/modules/CircularGradient/info.json]: And we could remove the "r" "g" "b" inputs here.

**Comment by HarshKhandeparkar** [comment]
Can you commit your new code?

**Comment by niravasher** [comment]
Updated my PR @HarshKhandeparkar  @jywarren

File: src/modules/CircularGradient/circularGradient.js [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+    var width = pixels.shape[0];
+    var height = pixels.shape[1];
+    if(height%2 != 0) {
+        height = height-1;
+    }
+    if(width%2 != 0) {
+        width = width-1;
+    }
+    var increment = width/2;
+    console.log(width);
```
Comment (jywarren) [inline src/modules/CircularGradient/circularGradient.js]: Oops, let's take out the log statements!

File: package.json [hunk 6]
```diff
@@ -36,6 +36,7 @@
     "image-sequencer-invert": "^1.0.0",
     "imagejs": "0.0.9",
     "imgareaselect": "git://github.com/jywarren/imgareaselect.git#v1.0.0-rc.2",
+    "infinite-gradients": "0.0.1",
```
Comment (jywarren) [inline package.json]: Are you using this? if not, let's take it back out, thanks!

File: src/modules/CircularGradient/circularGradient.js [hunk 7]
```diff
@@ -0,0 +1,37 @@
+module.exports = exports = function(pixels, options) {
+
+    let defaults = require('./../../util/getDefaults.js')(require('./info.json'));
+
+    let fillColor = options.fillColor || defaults.fillColor;
```
Comment (jywarren) [inline src/modules/CircularGradient/circularGradient.js]: OK so here we can just set `fillColor` to white, so we can remove all the fillColor code, make sense? we only want monochrome for now.

File: src/modules/CircularGradient/circularGradient.js [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+    fillColor = fillColor.split(" ")
+
+    for (var i = 0 ; i < width ; i++) {
+        for (var j = 0 ; j < height ; j++) {
+            var distX = Math.abs(options.centreX - i);
+            var distY = Math.abs(options.centreY - j);
+            var distance = Math.sqrt(Math.pow(distX, 2) + Math.pow(distY, 2));
+            var brightness = 255 * (distance / (2*width));
+            if (distance-1 > width) {distance = width;}
+            pixels.set(i, j, 0, fillColor[0]*1.5*distance/width);
```
Comment (jywarren) [inline src/modules/CircularGradient/circularGradient.js]: Right so on these three lines, we can just remove the `fillcolor[.]` parts, and it'll fill evenly with white.

File: src/modules/CircularGradient/info.json [hunk 9]
```diff
@@ -0,0 +1,25 @@
+{
+    "name": "Circular Gradient",
+    "description": "Gives a circular gradient of the image",
+    "inputs": {
+        "fillColor": {
```
Comment (jywarren) [inline src/modules/CircularGradient/info.json]: And let's remove this parameter here, we'll be able to do all color work in the next module!

File: src/modules/Gradient/Module.js [hunk 10]
```diff
@@ -4,52 +4,29 @@ module.exports = function Invert(options, UI) {
...
-            });
-
-            r.on("end", function() {
-                var data = Buffer.concat(chunks, totalLength).toString("base64");
-                var datauri = "data:image/" + input.format + ";base64," + data;
-                output(input.image, datauri, input.format);
-                callback();
-            });
-        });
+        function extraManipulation(pixels) {
```
Comment (jywarren) [inline src/modules/Gradient/Module.js]: Uh oh, would you mind leaving the existing Gradient module as it is for the time being? We can work on it in a follow-up step, so lets keep this PR simpler. Thanks!

File: src/modules/Gradient/info.json [hunk 11]
```diff
@@ -1,6 +1,13 @@
 {
     "name": "Gradient",
     "description": "Gives a gradient of the image",
-    "inputs": {},
+    "inputs": {
```
Comment (jywarren) [inline src/modules/Gradient/info.json]: Oh i see! OK, well, we can leave this in if you like, but perhaps in the future it's best to do only one concrete thing in each PR, so we have less to review and it's more self-contained. Make sense? Thank you!

Comment (jywarren) [reply src/modules/Gradient/info.json]: This is a cool addition to this module though 😄

File: src/modules/Gradient/Module.js (hunk 10 above)
Comment (niravasher) [reply src/modules/Gradient/Module.js]: Hey, but I added the feature for direction of gradient in this. Tell me of you if I should still send a PR without this and update this in the next PR?

File: src/modules/Gradient/info.json (hunk 11 above)
Comment (niravasher) [reply src/modules/Gradient/info.json]: Hehe, this is funny. Will take care for such kind of changes from the next time. 😄 😄

**Comment by niravasher** [comment]
Pushed all new suggested changes to this. Why am I still getting those conflicts? @HarshKhandeparkar @jywarren

**Comment by Divy123** [comment]
There are conflicts due to some other PRs already merged after your last rebase.
So you  can again make a rebase but lets work on the reviews by the reviewers

File: src/modules/CircularGradient/circularGradient.js [hunk 12]
```diff
@@ -0,0 +1,37 @@
...
+    console.log(height);
+
+    fillColor = fillColor.split(" ")
+
+    for (var i = 0 ; i < width ; i++) {
+        for (var j = 0 ; j < height ; j++) {
+            var distX = Math.abs(options.centreX - i);
+            var distY = Math.abs(options.centreY - j);
+            var distance = Math.sqrt(Math.pow(distX, 2) + Math.pow(distY, 2));
+            var brightness = 255 * (distance / (2*width));
```
Comment (Divy123) [inline src/modules/CircularGradient/circularGradient.js]: I think brightness is unused here.
So this can be removed.

**Comment by niravasher** [comment]
Ill make all the changes suggested by you @Divy123 , I am just not able to get rid of these merge conflicts. I am making unwanted commits.

**Comment by Divy123** [comment]
@niravasher what can be done here is you can go one commit behind on this.
I think some wrong files have changed here due to rebase done incorrectly.
You can go behind one or two commits back when there was only your changes and I suggest there is no need to rebase until you are done with your work completely.
It should be the last step.
Please go a commit behind to that version where there were only your changes.

**Comment by Divy123** [comment]
We will see the merge conflicts once you are on the rght version.

**Comment by jywarren** [comment]
Hi! I spent some time to extract /only/ the circular gradient work from here, and make just one commit. @niravasher it may be helpful to read through some guides on Git workflow -- it can get pretty complex, but it's worth it and I know you'll be an expert in no time 😄 👍 

http://publiclab.org/developers#Resources has some good links, and https://dev.to has lots of nice guides too!

I'll make some suggestions now in the simplified commit.

File: src/modules/CircularGradient/circularGradient.js [hunk 13]
```diff
@@ -0,0 +1,37 @@
...
+            if (distance-1 > width) {distance = width;}
+            pixels.set(i, j, 0, fillColor[0]*1.5*distance/width);
+            pixels.set(i, j, 1, fillColor[1]*1.5*distance/width);
```
Comment (jywarren) [inline src/modules/CircularGradient/circularGradient.js]:
```suggestion
            pixels.set(i, j, 1, 255 * distance / width);
```

File: src/modules/CircularGradient/circularGradient.js [hunk 14]
```diff
@@ -0,0 +1,37 @@
...
+            pixels.set(i, j, 2, fillColor[2]*1.5*distance/width);
```
Comment (jywarren) [inline src/modules/CircularGradient/circularGradient.js]:
```suggestion
            pixels.set(i, j, 2, 255 * distance / width);
```

File: src/modules/CircularGradient/info.json (hunk 9 above)
Comment (jywarren) [reply src/modules/CircularGradient/info.json]: Just reminding that we should remove this whole parameter, down to line 10!

**Comment by niravasher** [comment]
I am deleting this repository and forking once again, will make a new pull request @jywarren New PR will have all the changes suggested by you

**Comment by HarshKhandeparkar** [comment]
Wait!

**Comment by HarshKhandeparkar** [comment]
You don't have to retork

**Comment by HarshKhandeparkar** [comment]
You can run `git reset --hard upstream/main` to make your current branch perfectly even with upstream/main.

You can run this command on your own main branch and then create a new branch from it

**Comment by HarshKhandeparkar** [comment]
Another pr for the same is open.

**Pull request closed without merging**